src/snowflake/snowpark/modin/plugin/_internal/join_utils.py

```diff
@@ -161,40 +204,31 @@ def join(
                 include mapping for index + data columns, ordering columns and row position column
                 if exists.
     """
-    assert len(left_on) == len(
-        right_on
-    ), "left_on and right_on must be of same length or both be None"
-    if join_key_coalesce_config is not None:
-        assert len(join_key_coalesce_config) == len(
-            left_on
-        ), "join_key_coalesce_config must be of same length as left_on and right_on"
     assert how in get_args(
         JoinTypeLit
     ), f"Invalid join type: {how}. Allowed values are {get_args(JoinTypeLit)}"
-
-    def assert_snowpark_pandas_types_match() -> None:
-        """If Snowpark pandas types do not match, then a ValueError will be raised."""
-        left_types = [
-            left.snowflake_quoted_identifier_to_snowpark_pandas_type.get(id, None)
-            for id in left_on
-        ]
-        right_types = [
-            right.snowflake_quoted_identifier_to_snowpark_pandas_type.get(id, None)
-            for id in right_on
-        ]
-        for i, (lt, rt) in enumerate(zip(left_types, right_types)):
-            if lt != rt:
-                left_on_id = left_on[i]
-                idx = left.data_column_snowflake_quoted_identifiers.index(left_on_id)
-                key = left.data_column_pandas_labels[idx]
-                lt = lt if lt is not None else left.get_snowflake_type(left_on_id)
-                rt = rt if rt is not None else right.get_snowflake_type(right_on[i])
-                raise ValueError(
-                    f"You are trying to merge on {type(lt).__name__} and {type(rt).__name__} columns for key '{key}'. "
-                    f"If you wish to proceed you should use pd.concat"
-                )
-
-    assert_snowpark_pandas_types_match()
+    if how == "asof":
+        assert (
+            left_match_col
```

Comment:
for asof join, we can provide both match col and join on column, what would happen if both on and match columns are provided? if we only have tested support for match column, let's error out if both are provided today

Reply:
We can have both `on` and `match` columns for ASOF Join (in Snowflake SQL + Snowpark Python APIs). The Python API does have tests for both though even though we don't use it in the pandas layer. 

I don't think we need to error out in pandas if both are provided, but can change it to do so if you think it's better.

Reply:
What i am not worried here is not whether we error it out or not, but more about that we are not handling it correctly. For example, if i provide bout match col and on, should the join key be both match_col and on columns, or just on columns, or just match cols?
if we support both, we should make sure we have test for it

Reply:
@sfc-gh-azhan @sfc-gh-yzou I added support in `pd.merge_asof` for `by`, `left_by`, and `right_by` with tests to show how `left_on` and `right_on` are used with ASOF Join since it wasn't much additional work and it's a good test case for this.